Update NPC Generator case study lesson to use Google Gemini (#14)
Ada-Developers-Academy/independent-ai-topics: working-with-openai/open-ai-api-in-projects.md

Our goals for this lesson are to:
- Integrate the `google-generativeai` library into a Python project.
- Use the `/gemini-1.5-flash:generateContent` text completion endpoint to generate content for our project. 

## Set Up the NPC Generator

This tutorial will walk you through how to integrate the Gemini API into your projects with the companion repo [NPC Generator](https://github.com/AdaGold/NPC-Generator). We'll start with our usual setup.

1. Fork the repo to your GitHub account and then clone it to your machine.
2. Create and run a virtual environment for the project.
3. Use `pip` to install the `requirements.txt`.

## Explore the NPC Generator repo

Take some time to explore the NPC Generator repo in its current state. For this project, we will be finishing an API for a role-playing style game. Our API creates NPCs (Non Playable Characters) that your player's character will interact with. 

> We currently have two models, "Character" and "Greeting". "Character" holds some basic information about a character and has a one to many relationship with "Greeting", which represents a stock phrase that particular character may say when interacting with the player. Each NPC can have multiple greetings but each greeting can only be connected to one character. We will use the Gemini API to generate those greetings which we will then store in our database.

Reviewer: ```suggestion
We currently have two models, "Character" and "Greeting". "Character" holds some basic information about a character and has a one to many relationship with "Greeting", which represents a stock phrase that a particular character may say when interacting with the player. Each NPC can have multiple greetings but each greeting can only be connected to one character. We will use the Gemini API to generate those greetings which we will then store in our database.
```

</details>
</br>

## Install the `google-generativeai` Library and Connect Your Secret Key

Before we write our request to the Gemini API within the project, we need to install the `google-generativeai` package. 

1. In your terminal, run `pip install -q -U google-generativeai`
2. Run `pip freeze > requirements.txt` to add the `google-generativeai` package to your requirements.
3. In your `.env` file, add the variable `GEMINI_API_KEY` and assign it the value of your secret key: 
    `GEMINI_API_KEY=<Your Secret Key>`
4. In the `character_routes.py` file, import the `google-generativeai` package and configure the Gemini API environment using the code below:
   
    ```python
    import google.generativeai as genai
    import os

    genai.configure(api_key=os.environ.get("GEMINI_API_KEY"))
    ```
   
   | <div style="min-width:300px;">Part of Code</div> | Description |
   | ------------- | ----------- | 
   | `import google.generativeai as genai` | Import the `google-generativeai` package and give it an alias of `genai`. |
   | `import os` | Import the `os` package so we can read our `GEMINI_API_KEY` from the `.env` file. |

>    | `genai.configure(api_key=...)` | This code creates a Gemini API environment with your secret key from the `.env` file. We will use `genai` to access the `gemini-1.5-flash` model and make requests to the `/gemini-1.5-flash:generateContent` endpoint. |

Author: ```suggestion
   | `genai.configure(api_key=...)` | This code configures the Gemini API environment with your secret key from the `.env` file. We will use the `genai` import to access the `gemini-1.5-flash` model and make requests to the `/gemini-1.5-flash:generateContent` endpoint. |
```

            "greeting" : greeting.greeting_text
        })
    
    return response
```
</details>
</br>

## Summary
In this lesson, we took a partially built API and expanded it to include AI generated content. This is just the tip of the iceberg when it comes to integrating the AI APIs into your code, so we encourage you as always to follow your curiosity and keep looking for new and exciting ways to bring AI into your projects!

> 

Reviewer: Not sure if this is necessary, but would it be helpful to include a callout along the lines below?

```suggestion
### !callout-info

## Other AI APIs and Python
As mentioned before, Gemini is just one of many AI APIs you can integrate into your Python projects. We encourage you to explore others as your time and resources allow. While they won't look exactly the same, the steps for integration should be similar. There will likely be a Python package you can install with functions that will allow you to designate a model, attach your API key, make a request and more! Make sure to follow the documentation for whichever API you end up using and try things out!

### !end-callout  
```

Author: I think this is great, thank you for adding it!

            "greeting" : greeting.greeting_text
        })
    
    return response
```
</details>
</br>

## Summary
In this lesson, we took a partially built API and expanded it to include AI generated content. This is just the tip of the iceberg when it comes to integrating the AI APIs into your code, so we encourage you as always to follow your curiosity and keep looking for new and exciting ways to bring AI into your projects!

> ### !callout-info

Author: ```suggestion

### !callout-info
```

            "greeting" : greeting.greeting_text
        })
    
    return response
```
</details>
</br>

## Summary
In this lesson, we took a partially built API and expanded it to include AI generated content. This is just the tip of the iceberg when it comes to integrating the AI APIs into your code, so we encourage you as always to follow your curiosity and keep looking for new and exciting ways to bring AI into your projects!

### !callout-info

## Other AI APIs and Python
As mentioned before, Gemini is just one of many AI APIs you can integrate into your Python projects. We encourage you to explore others as your time and resources allow. While they won't look exactly the same, the steps for integration should be similar. There will likely be a Python package you can install with functions that will allow you to designate a model, attach your API key, make a request and more! Make sure to follow the documentation for whichever API you end up using and try things out!

> ### !end-callout  
> ## Check For Understanding

Author: ```suggestion
### !end-callout  

## Check For Understanding
```
